kinesis/src/main/scala/akka/stream/alpakka/kinesis/impl/KinesisSchedulerSourceStage.scala

```diff
@@ -0,0 +1,105 @@
+/*
+ * Copyright (C) 2016-2019 Lightbend Inc. <http://www.lightbend.com>
+ */
+
+package akka.stream.alpakka.kinesis.impl
+
+import java.util.concurrent.Semaphore
+
+import akka.actor.ActorRef
+import akka.annotation.InternalApi
+import akka.stream.alpakka.kinesis.KinesisSchedulerErrors.SchedulerUnexpectedShutdown
+import akka.stream.alpakka.kinesis.{CommittableRecord, KinesisSchedulerSourceSettings}
+import akka.stream.stage.GraphStageLogic.StageActor
+import akka.stream.stage._
+import akka.stream.{Attributes, Outlet, SourceShape}
+import software.amazon.kinesis.coordinator.Scheduler
+import software.amazon.kinesis.processor.{ShardRecordProcessor, ShardRecordProcessorFactory}
+
+import scala.collection.mutable
+import scala.concurrent.{ExecutionContext, Future, Promise}
+import scala.util.{Failure, Success, Try}
+
+/**
+ * Internal API
+ */
+@InternalApi
+private[kinesis] object KinesisSchedulerSourceStage {
+
+  final case class NewRecord(cr: CommittableRecord)
+  final case object Pump
+  final case object Complete
+  final case class SchedulerShutdown(result: Try[_])
+
+}
+
+/**
+ * Internal API
+ */
+@InternalApi
+private[kinesis] class KinesisSchedulerSourceStage(
+    settings: KinesisSchedulerSourceSettings,
+    schedulerBuilder: ShardRecordProcessorFactory => Scheduler
+)(implicit ec: ExecutionContext)
+    extends GraphStageWithMaterializedValue[SourceShape[CommittableRecord], Future[Scheduler]] {
+
+  private val out = Outlet[CommittableRecord]("Records")
+  override def shape: SourceShape[CommittableRecord] = new SourceShape[CommittableRecord](out)
+
+  override def createLogicAndMaterializedValue(
+      inheritedAttributes: Attributes
+  ): (GraphStageLogic, Future[Scheduler]) = {
+    val matValue = Promise[Scheduler]()
+    new Logic(matValue) -> matValue.future
+  }
+
+  final class Logic(matValue: Promise[Scheduler]) extends GraphStageLogic(shape) with StageLogging with OutHandler {
+    setHandler(out, this)
+
+    import settings._
+    import KinesisSchedulerSourceStage._
+
+    // We're transmitting backpressure from the Outlet to the Scheduler using a semaphore instance
+    // semaphore.acquire ~> callback ~> push downstream ~> semaphore.release
+    private[this] val semaphore = new Semaphore(bufferSize)
+    private[this] val buffer = mutable.Queue.empty[CommittableRecord]
+    private[this] var schedulerOpt: Option[Scheduler] = None
+    private[this] var self: StageActor = _
+
+    override def preStart(): Unit = {
+      self = getStageActor(awaitingRecords)
+      val newRecordCallback: CommittableRecord => Unit = {
+        semaphore.tryAcquire(backpressureTimeout.length, backpressureTimeout.unit)
+        self.ref ! NewRecord(_)
```

Comment:
This is OK, but most other Alpakka connectors use async callbacks instead.

Reply:
Please check the async callbacks logic now, is a bit less fair than before as I turned some of the intra-stage calls synchronous. I expect this to have higher performance. 

This can still be changed to everything asynchronous to maximize fairness.